azavea/grout
# Fix regression in JSONB filters

# Overview

This PR fixes a regression that caused JSONB filters to never return Records, even for valid queries.

## Notes

- See https://github.com/azavea/grout-2018-fellowship/issues/47#issuecomment-410057197 for a detailed trace of the source of the bug. TL;DR: the application code never registered the custom JSON lookup from the `lookups` module, so JSONField silently defaulted to the built-in filter behavior. On the other hand, the tests were configured in such a way that the `lookups` module got imported during test setup, causing tests to pass while the bug persisted in production.

## Testing instructions

Unit tests for this bug would be a real pain to write, since one of the key features of the bug is that the lookups get loaded automatically in the tests but not in the application code.

Instead, you can manually test that the changes work by following these steps:

1. First, **remove the test module that loads the custom lookups**:
    1. Edit `tests/settings_test.py` and remove line 29, thereby uninstalling `jsonb_field_testing` from `INSTALLED_APPS`
    2. Remove the `jsonb_field_testing` test module: `rm -Rf tests/jsonb_field_testing`
2. Next, **edit the application code to revert the changes made in this branch** in order to confirm that the bug exists:
    1. Edit `grout/apps.py` and remove lines 8-11, thereby failing to import the `lookups` module
    1. Run tests with `./scripts/test app py36-django20`  (you should see two failures)
3. **Add the changes from this branch back in** so you can confirm that the tests pass again:
    1. `git checkout grout/apps.py`
    2. `./scripts/test app py36-django20` (all tests should pass)
4. **Add the test module that you removed back in** and confirm that all other tests are passing:
    1. `git checkout tests/jsonb_field_testing`
    2. `git checkout tests/settings_test.py`
    3. `./scripts/test app py36-django20` (all tests should pass)

Closes https://github.com/azavea/grout-2018-fellowship/issues/47.
